---
title: "tSNE for scRNA-seq"
author: "Chela"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: hide
---


```{r check_r_version, results="asis"}
seed = 93
cat(paste0("<div class=\"alert alert-info\">
<strong>NOTE:</strong> The results of this script change depending on the version of R and Bioconductor (this affects the data normalisation). The original version (Chela's) was done with R version 3.5.1 and Bioconductor 3.8. Very similar results are possible with more recent version, although we change the seed for the t-SNE to obtain a plot that is visually quite similar.
<p>Another difference is in removing the outliers. In the new version we do not remove them. This is because the (a) data have been filtered already and (b)these filters have a high tendency to detect quiescent cells.</div>"), fill = T)
```

## Introduction

This analysis (FKA marioni.Rmd) is based on a tutorial from Marioni's group, initially run by Chela and later modified by Javier.

**NOTE**: The results of this script change depending on the version of R and Bioconductor. Please use R version 3.5.1 and Bioconductor 3.8 if you want to reproducee the t-SNE.

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = TRUE,
  warning = FALSE)
# knitr::opts_chunk$set(cache = TRUE, autodep = TRUE)
```

```{r, message = FALSE}
library(scater)
library(pheatmap)
library(biomaRt)
library(tidyverse)
library(patchwork)
library(ggbeeswarm)
library(dynamicTreeCut)
data("paper_palette")
```

## Data

For this analysis we use the data in `FinalPaper_SingleCellExperiment_PDX.23072018.RDS` and reannotate the genes with the table from e!84.

PRS cells here are defined from the TSCAN pseudotime, i.e. untreated cells from clusters 6, 7 and 9, but we will defined other PRS sets later on.

```{r}
# setup data
# pdx_sce_filename <- "results/FinalPaper_SingleCellExperiment_PDX_chela_edit.10_10_19.RDS"
# pdx_sce_filename <- "data/FinalPaper_SingleCellExperiment_PDX.23072018.RDS"
# sce <- readRDS(pdx_sce_filename)
data("sce.pdx")

# incorporate gene based annotation
# ensembl = useMart("ensembl",dataset="hsapiens_gene_ensembl")
# my_df <- getBM(attributes = c('ensembl_gene_id', 'hgnc_symbol', 'chromosome_name','start_position', 'end_position',
#                               'band'),
#                filters = 'ensembl_gene_id',
#                rownames(sce),
#                mart = ensembl)
# # this gene is duplicated because it defines two genes CCL3L1 and CCL3L3 
# my_df <- my_df[-14738,]

load("data/t2g.e84.RData")
my_df<-left_join(rownames_to_column(data.frame(rowData(sce.pdx)),"ensembl_gene_id"),
                 t2g.e84 %>%
                   dplyr::select(ensembl_gene_id = ens_gene,
                          hgnc_symbol = ext_gene,
                          chromosome_name) %>%
                   unique(),
                 by = "ensembl_gene_id")

my_df$hgnc_symbol <- case_when(is.na(my_df$hgnc_symbol) ~ my_df$ensembl_gene_id,
                     my_df$hgnc_symbol == "" ~ my_df$ensembl_gene_id, 
                     TRUE ~ as.character(my_df$hgnc_symbol))

rownames(my_df) <- my_df$ensembl_gene_id
rowData(sce.pdx) <- my_df

mito <- which(rowData(sce.pdx)$chromosome_name == "MT")
sce.pdx <- calculateQCMetrics(sce.pdx, feature_controls=list(Mt=mito))
whichTSCANFreddies <- sce.pdx$PRS_2021 == "PRS"

sce.pdx$isPRS <- "No"
sce.pdx$isPRS[whichTSCANFreddies] <- "Yes"

sce.pdx$isPRS <- factor(sce.pdx$isPRS, levels = c("No", "Yes"), labels = c("No", "Yes"))

## Note: this is probably defined in the FinalPaper_SingleCellExperiment_PDX_chela_edit.10_10_19.RDS
## dataset, but needs to be added here
multiplot(
    plotColData(sce.pdx, y = "total_counts", x = "PRS_2021"),
    plotColData(sce.pdx, y = "total_features_by_counts", x = "PRS_2021"),
    plotColData(sce.pdx, y = "pct_counts_Mt", x = "PRS_2021"),
    cols=3)
```


```{r}
par(mfrow=c(1,2))
plot(sce.pdx$total_features_by_counts, sce.pdx$total_counts/1e6, xlab="Number of expressed genes",
    ylab="Library size (millions)")
plot(sce.pdx$total_features_by_counts, sce.pdx$pct_counts_Mt, xlab="Number of expressed genes",
    ylab="Mitochondrial proportion (%)")
```

### Average count for each gene

Another useful metric is the average count for each gene, computed across all cells in the dataset. We calculate this using the calcAverage() function, which also performs  some adjustment for library size differences between cells. We typically observe a  peak of moderately expressed genes following a plateau of lowly expressed genes.

```{r}
ave.counts <- calcAverage(sce.pdx, use_size_factors=FALSE)
hist(log10(ave.counts), breaks=100, main="", col="grey80", 
    xlab=expression(Log[10]~"average count"))
```

### Number of cells that express each gene

We also examine the number of cells that express each gene. This is closely related  to the average count for most genes, as expression in many cells will result in a  higher average (Figure 6). Genes expressed in very few cells are often uninteresting  as they are driven by amplification artifacts (though they may also also arise from rare populations).

```{r}
num.cells <- nexprs(sce.pdx, byrow=TRUE)
smoothScatter(log10(ave.counts), num.cells, ylab="Number of cells", 
    xlab=expression(Log[10]~"average count"))
```

### Normalization

*Read counts are subject to differences in capture efficiency and sequencing depth between cells (Stegle, Teichmann, and Marioni 2015). Normalization is required to eliminate these cell-specific biases prior to downstream quantitative analyses. This is often done by assuming that most genes are not differentially expressed (DE) between cells. Any systematic difference in count size across the non-DE majority of genes between two cells is assumed to represent bias and is removed by scaling. More specifically, “size factors” are calculated that represent the extent to which counts should be scaled in each library.*

*Size factors can be computed with several different approaches, e.g., using the estimateSizeFactorsFromMatrix function in the DESeq2 package (Anders and Huber 2010; Love, Huber, and Anders 2014), or with the calcNormFactors function (Robinson and Oshlack 2010) in the edgeR package. However, single-cell data can be problematic for these bulk data-based methods due to the dominance of low and zero counts. To overcomethis, we pool counts from many cells to increase the size of the counts for accurate size factor estimation (Lun, Bach, and Marioni 2016). Pool-based size factors are then “deconvolved” into cell-based factors for normalization of each cell’s expression profile.*

```{r}
library(scran)

sce.pdx <- computeSumFactors(sce.pdx)
summary(sizeFactors(sce.pdx))
sce.pdx$group4 <- factor(sce.pdx$group4)


# norm.factors <- edgeR::calcNormFactors(assay(sce.pdx, "counts"), method = "TMM")
# tmm.size.factors <- norm.factors * colSums(assay(sce.pdx, "counts"))
# sizeFactors(sce.pdx) <- tmm.size.factors
# 
# 
plot(sce.pdx$total_counts/1e6, sizeFactors(sce.pdx), log="xy",
    xlab="Library size (millions)", ylab="Size factor",
    col=c("red", "black")[sce.pdx$group4], pch=16)
legend("bottomright", col=c("red", "black"), pch=16, cex=1.2,
    legend=levels(sce.pdx$group4))

# sce.pdx <- normalize(sce.pdx)

sce.pdx <- scater::normalize(sce.pdx)
# sce.pdx <- scater::normalizeCounts(sce.pdx)
```

### Variance

```{r}
rownames(sce.pdx) <- rowData(sce.pdx)$hgnc_symbol
var.fit <- trendVar(sce.pdx, parametric=TRUE, block=sce.pdx$group4,
    loess.args=list(span=0.3), use.spikes = FALSE)
var.out <- decomposeVar(sce.pdx, var.fit)


plot(var.out$mean, var.out$total, pch=16, cex=0.6, xlab="Mean log-expression", 
    ylab="Variance of log-expression")
curve(var.fit$trend(x), col="dodgerblue", lwd=2, add=TRUE)
rownames(var.out) <- rownames(sce.pdx)
chosen.genes <- order(var.out$bio, decreasing=TRUE)[1:10]
```

Expression of the top 10 most variable genes:

```{r}
plotExpression(sce.pdx, features=rownames(var.out)[chosen.genes]) 
```

## Dimensionality reduction

### PCA

```{r}
sce.pdx <- denoisePCA(sce.pdx, technical=var.out)
plotReducedDim(sce.pdx, use_dimred="PCA", ncomponents=3, 
    colour_by="group4") 
```


### t-SNE

T-SNE analysis with perplexity = 20 (although aestetically similar results can be achieved with perplexity = 10 as long as the thresholds on t-SNE map are adjusted).



```{r fig.width = 3.7, fig.height = 3.2}
set.seed(seed)
sce.pdx <- runTSNE(sce.pdx, dimred="PCA", perplexity = 20)
out20 <- plotTSNE(sce.pdx, run_args=list(dimred="PCA", perplexity=20),
    colour_by="PRS_2021")

reducedDimNames(sce.pdx)

ggplot(out20$data) +
  geom_point(aes(x = X, y = Y, col = colour_by), size = 0.8) +
  scale_color_manual(values = paper_palette, name = "group") +
  labs(x = "TSNE 1", y = "TSNE 2") +
  theme_classic() +
  theme(text = element_text(size = 10),
        legend.position = c(0.95, 0.05),
        legend.justification = c(1, 0),
        legend.title = element_blank(),
        legend.key.size = unit(0.03, "npc"),
        legend.margin = margin(0, 5, 5, 5),
        legend.background = element_rect(fill = NA, color = "black", size = 0.3))

ggsave("output/figures/Fig8D_tSNE_plot.pdf", device = "pdf", width = 3.7, height = 3.2)

wb <- openxlsx::createWorkbook()
ws_name <- "Fig8E PRS signature"
ws <- openxlsx::addWorksheet(wb, ws_name)


excel_data <- colData(sce.pdx) %>%
  as_tibble(rownames = "Cell") %>%
  dplyr::select(Cell, Group = PRS_2021) %>%
  left_join(reducedDim(sce.pdx, "TSNE") %>%
              as_tibble(rownames = "Cell") %>%
              dplyr::rename(TSNE1 = V1, TSNE2 = V2),
            by = "Cell")

openxlsx::writeDataTable(wb, sheet = ws, x = excel_data,
               rowNames = F, tableStyle = "none", withFilter = F)
openxlsx::setColWidths(wb, ws, cols = 1:ncol(excel_data), widths = "auto")
openxlsx::addStyle(wb, ws, rows = 1, cols = 1:ncol(excel_data),
                   style = openxlsx::createStyle(textDecoration = "bold"))
openxlsx::saveWorkbook(wb, "output/scRNAseq_data_Fig8D.xlsx", overwrite = TRUE)
```


## Defining PRS from the tSNE plot

### cell clusters using PCA

In here, we cluster the cells using the ward.D2 hierarchical clustering algorithm and the euclidead distance metric on the PCA reduced dimensions.

Note: The clusters are then used for the next plot and for the definition of the final set of PRS (based on the heatmap, where these are limited to cluster 1).

*The denoised log-expression values are used to cluster cells into putative
subpopulations. Specifically, we perform hierarchical clustering on the
Euclidean distances between cells, using Ward’s criterion to minimize the
total variance within each cluster. This yields a dendrogram that groups
together cells with similar expression patterns across the chosen genes.*

As can be shown in the TSNE plot colored by cluster, this produces 6 clusters.

```{r}
pcs <- reducedDim(sce.pdx, "PCA")
my.dist <- dist(pcs)
my.tree <- hclust(my.dist, method="ward.D2")

my.clusters <- unname(cutreeDynamic(my.tree, distM=as.matrix(my.dist), 
    minClusterSize=20, verbose=0))

sce.pdx$cluster <- factor(my.clusters)
plotTSNE(sce.pdx, colour_by="cluster")
plotTSNE(sce.pdx, colour_by="PRS_2021")

# plotUMAP(sce.pdx, colour_by="cluster")
# plotUMAP(sce.pdx, colour_by="group")
# plotUMAP(sce.pdx, colour_by="PRS_2021")
# set.seed(100)
# plotUMAP(sce.pdx, colour_by="cluster")
# set.seed(200)
# plotUMAP(sce.pdx, colour_by="cluster")
# set.seed(300)
# plotUMAP(sce.pdx, colour_by="cluster")
# set.seed(400)
# plotUMAP(sce.pdx, colour_by="cluster")
```

*We check the separatedness of the clusters using the silhouette width 
(Figure 14). Cells with large positive silhouette widths are closer to
other cells in the same cluster than to cells in different clusters.
Conversely, cells with negative widths are closer to other clusters than
to other cells in the cluster to which it was assigned. Each cluster would
ideally contain many cells with large positive widths, indicating that 
it is well-separated from other clusters.*

```{r}
library(cluster)
clust.col <- scater:::.get_palette("tableau10medium") # hidden scater colours
sil <- silhouette(my.clusters, dist = my.dist)
sil.cols <- clust.col[ifelse(sil[,3] > 0, sil[,1], sil[,2])]
sil.cols <- sil.cols[order(-sil[,1], sil[,3])]
plot(sil, main = paste(length(unique(my.clusters)), "clusters"), 
    border=sil.cols, col=sil.cols, do.col.sort=FALSE) 
```

### Defining PRS from the t-SNE plot

We now define the prsTSNE cells as the untreated cells that appear in cluster 1 (i.e. with an X coordinate < 0 and Y < -10  in the t-SNE plot)

```{r}
my_TSNEplot_by_cluster <- plotTSNE(sce.pdx, colour_by="cluster")
my_TSNEplot_by_group <- plotTSNE(sce.pdx, colour_by="group")
colData(sce.pdx)$cluster_and_prsTSNE <- case_when(
  my_TSNEplot_by_group$data$colour_by == "Untreated" &
    my_TSNEplot_by_group$data$X < 0 &
    my_TSNEplot_by_group$data$Y < -10 ~ "prsTSNE",
  TRUE ~ as.character(my_TSNEplot_by_cluster$data$colour_by))

plotTSNE(sce.pdx, colour_by="cluster_and_prsTSNE")

# colnames(sce.pdx)[which(colData(sce.pdx)$cluster_and_prsTSNE == "prsTSNE")]
# [1] "Untreated_VT14_N702_S503" "Untreated_VT14_N704_S504" "Untreated_VT14_N704_S505"
# [4] "Untreated_VT14_N707_S505" "Untreated_VT14_N707_S507" "Untreated_VT18_N703_S506"
```


### Heatmaps

Showing the location of the PRS cells in a Heatmap. This shows that a subset of the PRS cells indeed cluster with the treated cells.

```{r fig.height=8,fig.width=8}
markers <- findMarkers(sce.pdx, my.clusters)
marker.set <- markers[["1"]]
top.markers <- rownames(marker.set)[marker.set$Top <= 10]

order <- enframe(colnames(sce.pdx), name = NULL, value = "cell") %>%
          cbind(group = sce.pdx$cluster) %>%
          pull(group) %>% order()

ann_col <- data.frame(group = as.factor(sce.pdx$PRS_2021),
                      cluster = as.factor(sce.pdx$cluster),
                      prsTSNE = as.factor(sce.pdx$cluster_and_prsTSNE == "prsTSNE"))
rownames(ann_col) <- colnames(sce.pdx)

heat.vals <- logcounts(sce.pdx[top.markers, ])
heat.vals <- heat.vals - rowMeans(heat.vals)
zlim <- c(-5, 5)
extreme <- max(abs(zlim))
zlim <- c(-extreme, extreme)
heat.vals[heat.vals < zlim[1]] <- zlim[1]
heat.vals[heat.vals > zlim[2]] <- zlim[2]

pheatmap(heat.vals[, order],
         scale = "none",
         cluster_cols = T,
         color <- eval(formals(pheatmap::pheatmap)$color, envir = environment(pheatmap::pheatmap)),
         annotation_col = ann_col,
         annotation_colors = list(group = paper_palette[c("Treated", "PRS", "Untreated")],
                                  # cluster = paperPallette[1:6],
                                  prsTSNE = c("TRUE" = "black", "FALSE" = "white")),
         annotation_names_col = F,
         fontsize_row = 10,
         show_colnames = F,
         legend = F)

pheatmap(heat.vals[, order],
         scale = "none",
         cluster_cols = F,
         color <- eval(formals(pheatmap::pheatmap)$color, envir = environment(pheatmap::pheatmap)),
         annotation_col = ann_col,
         annotation_colors = list(group = paper_palette[c("Treated", "PRS", "Untreated")],
                                  prsTSNE = c("TRUE" = "black", "FALSE" = "white")),
         annotation_names_col = F,
         fontsize_row = 10,
         show_colnames = F,
         legend = F)

prs_defined_by_heatmap <- colData(sce.pdx)$cluster == 1 & colData(sce.pdx)$cluster_and_prsTSNE == "prsTSNE"
prs_defined_by_heatmap <- case_when(prs_defined_by_heatmap == FALSE ~ as.character(sce.pdx$group4),
                                    prs_defined_by_heatmap == TRUE ~ "PRS")

ann_col$prs_defined_by_heatmap <- prs_defined_by_heatmap

t1 <- pheatmap(heat.vals[, order],
         scale = "none",
         cluster_cols = F,
         color <- eval(formals(pheatmap::pheatmap)$color, envir = environment(pheatmap::pheatmap)),
         annotation_col = ann_col[c("prs_defined_by_heatmap")],
         annotation_colors = list(prs_defined_by_heatmap =
                                    paper_palette[c("Treated", "PRS", "Untreated")]),
         annotation_names_col = F,
         treeheight_row = 20,
         fontsize_row = 10,
         show_colnames = F,
         legend = F,
         silent = T)

g <- wrap_elements(t1$gtable)
print(g)

ggsave("output/figures/FigS8f_heatmap_PRS.pdf", device = "pdf")

colData(sce.pdx)$prs_defined_by_heatmap <- prs_defined_by_heatmap

excel_data <- ann_col %>%
  as_tibble(rownames = "Cell") %>%
  dplyr::select(Cell, Group = prs_defined_by_heatmap) %>%
  left_join(t(heat.vals[, order]) %>%
              as_tibble(rownames = "Cell"),
            by = "Cell")

wb <- openxlsx::createWorkbook()
ws <- openxlsx::addWorksheet(wb, "ExtFigS8f heatmap")

openxlsx::writeDataTable(wb, sheet = ws, x = excel_data,
               rowNames = F, tableStyle = "none", withFilter = F)
openxlsx::setColWidths(wb, ws, cols = 1:ncol(excel_data), widths = "auto")
openxlsx::addStyle(wb, ws, rows = 1, cols = 1:ncol(excel_data),
                   style = openxlsx::createStyle(textDecoration = "bold"))
openxlsx::addStyle(wb, ws, rows = 1:nrow(excel_data) + 1, cols = 1,
                   style = openxlsx::createStyle(textDecoration = "bold"))

openxlsx::saveWorkbook(wb, "output/scRNAseq_data_FigS8f.xlsx", overwrite = TRUE)
```

<!-- Same plot but a tad shorter -->

```{r fig.height=8,fig.width=12, eval = F}
pheatmap(heat.vals[, order],
         scale = "none",
         cluster_cols = F,
         color <- eval(formals(pheatmap::pheatmap)$color, envir = environment(pheatmap::pheatmap)),
         annotation_col = ann_col,
         annotation_colors = list(group = paper_palette[c("Treated", "PRS", "Untreated")]),
         annotation_names_col = F,
         fontsize_row = 10,
         show_colnames = F,
         legend = F)
```

## Comparison of different PRS cell definitions

We now have 3 definitions of PRS cells:

* PRS from the TSCAN pseudotime analyiss
* PRS from the clustering and mapping on the tSNE (6 cells; prsTSNE)
* PRS from the clustering and mapping on the tSNE after removing the PRS in the different cluster (5 cells; prs_defined_by_heatmap)

```{r}
sce.pdx$chemo_untreated_prs2 <- 
  ifelse(sce.pdx$cluster_and_prsTSNE == "prsTSNE", "PRS", as.character(sce.pdx$group4))

sce.pdx$chemo_untreated_prs3 <- 
  ifelse(sce.pdx$prs_defined_by_heatmap == "PRS", "PRS", as.character(sce.pdx$group4))

g1 <- colData(sce.pdx) %>%
  as_tibble(rownames = "Cell") %>%
  ggplot() +
  geom_violin(aes(x = PRS_2021, y = pct_counts_Mt)) +
  ggbeeswarm::geom_quasirandom(aes(x = PRS_2021, y = pct_counts_Mt,
                                   col = PRS_2021),
                               size = 1) +
  scale_color_manual(values = paper_palette, guide = "none") +
  labs(x = NULL, y = "Percentage MT", title = "TSCAN classification") +
  theme_bw() +
  theme(panel.grid = element_blank())

g2 <- colData(sce.pdx) %>%
  as_tibble(rownames = "Cell") %>%
  ggplot() +
  geom_violin(aes(x = chemo_untreated_prs2, y = pct_counts_Mt)) +
  ggbeeswarm::geom_quasirandom(aes(x = chemo_untreated_prs2, y = pct_counts_Mt,
                                   col = chemo_untreated_prs2),
                               size = 1) +
  scale_color_manual(values = paper_palette, guide = "none") +
  labs(x = NULL, y = "Percentage MT", title = "TSNE classification") +
  theme_bw() +
  theme(panel.grid = element_blank())

print(g1 + g2)

ggsave("output/figures/FigS8g_violin_PRS.pdf",
       device = "pdf", width = 4, height = 2, scale = 1.3)
# multiplot(
#     plotColData(sce.pdx, y = "pct_counts_Mt", x = "PRS_2021"),
#     plotColData(sce.pdx, y = "pct_counts_Mt", x = "chemo_untreated_prs2"),
#     cols=2)

excel_data <- colData(sce.pdx) %>%
  as_tibble(rownames = "Cell") %>%
  dplyr::select(Cell,
                `TSCAN Group` = PRS_2021,
                `TSNE Group` = chemo_untreated_prs2,
                pct_counts_Mt)

wb <- openxlsx::createWorkbook()
ws <- openxlsx::addWorksheet(wb, "ExtFigS8g PRS classif")

openxlsx::writeDataTable(wb, sheet = ws, x = excel_data,
               rowNames = F, tableStyle = "none", withFilter = F)
openxlsx::setColWidths(wb, ws, cols = 1:ncol(excel_data), widths = "auto")
openxlsx::addStyle(wb, ws, rows = 1, cols = 1:ncol(excel_data),
                   style = openxlsx::createStyle(textDecoration = "bold"))
openxlsx::addStyle(wb, ws, rows = 1:nrow(excel_data) + 1, cols = 1,
                   style = openxlsx::createStyle(textDecoration = "bold"))

openxlsx::saveWorkbook(wb, "output/scRNAseq_data_FigS8g.xlsx", overwrite = TRUE)
```

